csaf-matching/src/commonMain/kotlin/io/github/csaf/sbom/matching/Matcher.kt

```diff
@@ -16,16 +16,99 @@
  */
 package io.github.csaf.sbom.matching
 
+import io.github.csaf.sbom.matching.cpe.Cpe
+import io.github.csaf.sbom.matching.cpe.parseCpe
+import io.github.csaf.sbom.matching.purl.Purl
 import io.github.csaf.sbom.schema.generated.Csaf
+import io.github.csaf.sbom.validation.tests.mapBranchesNotNull
+import protobom.protobom.Document
+import protobom.protobom.SoftwareIdentifierType
 
 /**
- * Matches the provided SBOM document with the CSAF document and determines whether they meet
- * specific criteria.
+ * Matcher for matching SBOM documents with provided CSAF documents.
  *
- * @param sbom The SBOM document represented by a protobom.protobom.Document instance.
- * @param doc The CSAF document to be matched against, represented by a Csaf instance.
- * @return A boolean value indicating whether the SBOM document matches the CSAF document.
+ * @property docs The CSAF documents used by this matcher.
+ * @property threshold The default threshold required for a match to be included.
  */
-fun match(sbom: protobom.protobom.Document, doc: Csaf): Boolean {
-    return true
+class Matcher(val docs: List<Csaf>, val threshold: Float = 0.5f) {
+    val purlMap = mutableMapOf<String, MutableSet<FastHash<Csaf>>>()
+    val cpeMap = mutableMapOf<Cpe, MutableSet<FastHash<Csaf>>>()
+
+    /**
+     * The constructor checks that the given threshold is within its bounds and then extracts all
+     * relevant information for comparison from the given CSAF documents for faster matching.
+     */
+    init {
+        require(threshold in 0.0..1.0) { "Threshold must be in the interval [0.0; 1.0]." }
+        docs.forEach { doc ->
+            // Cache all canonicalized PURLs
+            doc.product_tree
```

Comment:
We need to match on the vulnerability's product status, not the product tree!

Reply:
Fixed